### PR TITLE
Rewrite the grammar once again

### LINKED ISSUE #74: Grammar defines trailing whitespace as part of lambda case statement
The `exp_lambda_case` node associated with the first definition includes the trailing white space:

Screenshot 2022-03-07 at 15 47 50

This results, for example, in incorrect deletions when operating based off the AST, with a deletion of the lambda case expression resulting in:

Screenshot 2022-03-07 at 15 50 06

### LINKED ISSUE #82: Comments following function included in function pattern
For functions with a `do` block, the comments following the function get included in the function, for example:

```
f = do a

-- | haddock
g = b
```

here the function pattern will include all of `f` and the doc comment of `g`. This isn't the case when there is no `do` block:

```
f = a

-- | haddock
g = b
```

in this case it works as I expected, only matching `f = a`.

I tested this out using the latest commit on the master branch, using the following tree sitter query:

```
(function rhs: (_) @function.inside) @function.around
```

(both captures end up including the doc comment)

### LINKED ISSUE #89: Instance with associated type, following TH top level splice, misparsed as function
I have done a run of executing tree-sitter-haskell against my employer Mercury's (closed source) codebase (350k LOC) and found a few parser bugs, which I have committed simplified testcases for on https://github.com/lf-/tree-sitter-haskell/tree/top-level-splices-oopsie. I am going to try to fix the ones that I think I can fix over the next few weeks.

---

I'm filing this incorrect parse as a bug because I am baffled by it and don't know how to fix it. My best guess is that for some reason the function parser gets greedy and ignores the semicolons?

The following test case misparses as a `function` node. Notably, if the `instance` is changed to have an associated value or such, it will parse correctly.

```
================================================================================
template haskell: top level splice without parens, but weird
================================================================================

someTemplateHaskell $(spliceOne)

instance SomeClass Something where
  type Assoc Something = ()

---
```

AST:

```
(haskell [0, 0] - [4, 0]
  (function [0, 0] - [3, 27]
    name: (variable [0, 0] - [0, 19])
    patterns: (patterns [0, 20] - [3, 22]
      (splice [0, 20] - [0, 32]
        (exp_parens [0, 21] - [0, 32]
          (exp_name [0, 22] - [0, 31]
            (variable [0, 22] - [0, 31]))))
      (pat_name [2, 0] - [2, 8]
        (variable [2, 0] - [2, 8]))
      (pat_name [2, 9] - [2, 18]
        (constructor [2, 9] - [2, 18]))
      (pat_name [2, 19] - [2, 28]
        (constructor [2, 19] - [2, 28]))
      (pat_name [2, 29] - [2, 34]
        (variable [2, 29] - [2, 34]))
      (pat_name [3, 2] - [3, 6]
        (variable [3, 2] - [3, 6]))
      (pat_name [3, 7] - [3, 12]
        (constructor [3, 7] - [3, 12]))
      (pat_name [3, 13] - [3, 22]
        (constructor [3, 13] - [3, 22])))
    rhs: (exp_literal [3, 25] - [3, 27]
      (con_unit [3, 25] - [3, 27]))))
```

### LINKED ISSUE #92: Misparse of explicit-braced code
I was reading some GHC sources when I found my highlighting was broken.

Here's a minimized example (which does compile with GHC):

```haskell
module OhNo where
lookupIdInScope :: Monad m => m ()
lookupIdInScope
  = do { case () of
           () -> do { return () } }
  where
    blah = ()
```

AST:
```
module: module [0, 7] - [0, 11]
where [0, 12] - [0, 17]
ERROR [1, 0] - [6, 13]
  pat_typed [1, 0] - [2, 15]
    pattern: pat_name [1, 0] - [1, 15]
      variable [1, 0] - [1, 15]
    type: context [1, 19] - [2, 15]
      constraint [1, 19] - [1, 26]
        class: class_name [1, 19] - [1, 24]
          type [1, 19] - [1, 24]
        type_name [1, 25] - [1, 26]
          type_variable [1, 25] - [1, 26]
      type_apply [1, 30] - [2, 15]
        type_name [1, 30] - [1, 31]
          type_variable [1, 30] - [1, 31]
        type_literal [1, 32] - [1, 34]
          con_unit [1, 32] - [1, 34]
        type_name [2, 0] - [2, 15]
          type_variable [2, 0] - [2, 15]
  exp_literal [3, 14] - [3, 16]
    con_unit [3, 14] - [3, 16]
  alt [4, 11] - [4, 33]
    pat_literal [4, 11] - [4, 13]
      con_unit [4, 11] - [4, 13]
    exp_do [4, 17] - [4, 33]
      stmt [4, 22] - [4, 31]
        exp_apply [4, 22] - [4, 31]
          exp_name [4, 22] - [4, 28]
            variable [4, 22] - [4, 28]
          exp_literal [4, 29] - [4, 31]
            con_unit [4, 29] - [4, 31]
  ERROR [4, 34] - [6, 10]
  pat_literal [6, 11] - [6, 13]
    con_unit [6, 11] - [6, 13]
```

### LINKED ISSUE #102: Include . from qualified modules and variables
Given the file:

```haskell
import System.Process
```

The parse tree is:

```
haskell (0, 0) - (1, 0)
  import (0, 0) - (0, 21)
    import (0, 0) - (0, 6) "import"
    qualified_module (0, 7) - (0, 21)
      module (0, 7) - (0, 13) "System"
      module (0, 14) - (0, 21) "Process"
```

There's no node for `.`, which makes life harder for AST tools like difftastic.

### LINKED ISSUE #108: exp_section_right not parsed when containing a hash
an expression like `(+ x) y` is [parsed correctly](https://felixroos.github.io/haskell-tree-sitter-playground/#KCsgeCkgeQ==), whereas `(# x) y` [yields an error](https://felixroos.github.io/haskell-tree-sitter-playground/#KCMgeCkgeQ==). Interestingly, [it works when prefixing a space](https://felixroos.github.io/haskell-tree-sitter-playground/#KCAjIHgpIHk=). I guess this must be a bug?

### LINKED ISSUE #116: Typed Template Haskell quotations / splices not handled correctly
GHC with `TemplateHaskell` allows typed quotations and splices, written
```
[|| x ||]
```
and
```
$$(foo bar)
```
or just
```
$$x
```

Looks like the tree-sitter parser currently handles this incorrectly, breaking syntax highlighting in rather substantial ways for files containing these constructs.

I looked briefly whether there's a very quick fix, but unfortunately it looks like this requires adapting the scanner.

### PR DESCRIPTION
**Pull request by tek**
I decided to redesign the scanner and parts of the grammar due to the large number of issues without obvious fix that
have accumulated over the years.
The new implementation provides a significant set of improvements, listed below.

Since the repo has [become uncloneable](https://github.com/tree-sitter/tree-sitter-haskell/issues/114) in some
situations, I would also prefer to change the workflow so that `parser.c` is only generated on a release branch, not for
every commit, like some other grammars do it.
However, since the current situation is already bad, it seems that the only way out would be to reset the history, which
would break consumers like nixpkgs who rely on being able to access older commits.
An alternative could be to use a new Github location, but that would also be awkward.
In any case, it's probably better to handle this later.

I've overhauled the tree structure a bit, mostly for higher-level nodes, since I found some parts to be lacking or badly
named.
For example, I added `header` / `imports` / `declarations` for the top-level structure.
Since I don't have much experience using the grammar via `tree-sitter` API directly, I've launched a [survey on the
Haskell discourse](https://discourse.haskell.org/t/survey-for-users-of-the-tree-sitter-haskell-grammar/9153) to get some more feedback, so I can use the opportunity of introducing breaking changes to
improve the grammar for users.

Not sure about the wasm artifact – is it still necessary to use that patch included in the `Makefile`?

I'd appreciate opinions and feedback!

Please take a look, @414owen @amaanq @wenkokke 

-----------------

* Parses the GHC codebase!

  I'm using a trimmed set of the source directories of the compiler and most core libraries in [this repo](https://github.com/tek/tsh-test-ghc).

  This used to break horribly in many files because explicit brace layouts weren't supported very well.

* Faster in most cases! Here are a few simple benchmarks to illustrate the difference, not to be taken _too_ seriously, using the test codebases in `test/libs`:

  Old: 
  ``` 
  effects: 32ms
  postgrest: 91ms
  ivory: 224ms
  polysemy: 84ms
  semantic: 1336ms
  haskell-language-server: 532ms
  flatparse: 45ms
   ```

  New:
  ```
  effects: 29ms
  postgrest: 64ms
  ivory: 178ms
  polysemy: 70ms
  semantic: 692ms
  haskell-language-server: 390ms
  flatparse: 36ms
  ```

  GHC's `compiler` directory takes 3000ms, but is among the fastest repos for per-line and per-character times! To get more detailed info (including new codebases I added, consisting mostly of core libraries), run `test/parse-libs`. I also added an interface for running `hyperfine`, exposed as a Nix app – execute `nix run .#bench-libs -- stm mtl transformers` with the desired set of libraries in `test/libs` or `test/libs/tsh-test-ghc/libraries`.

* Smaller size of the shared object.

  `tree-sitter generate` produces a `haskell.so` with a size of 4.4MB for the old grammar, and 3.0MB for the new one.

* Smaller size of the generated `parser.c`: 24MB -> 14MB.

* Significantly faster time to generate, and slightly faster build.

  On my machine, generation takes 9.34s vs 2.85s, and compiling takes 3.75s vs 3.33s.

* Smaller size of `parser.c`, from 24MB down to 14MB.

* All terminals now have proper text nodes when possible, like the `.` in modules. Fixes #102, #107, #115 (partially?).

* Semicolons are now forced after newlines even if the current parse state doesn't allow them, to fail alternative interpretations in GLR conflicts that sometimes produced top-level expression splices for valid (and invalid) code. Fixes #89, #105, #111.

* Comments aren't pulled into preceding layouts anymore. Fixes #82, #109. (Can probably still be improved with a few heuristics for e.g. postfix haddock)

* Similarly, whitespace is kept out of layout-related nodes as much as possible. Fixes #74.

* Hashes can now be operators in all situations, without sacrificing unboxed tuples. Fixes #108.

* Expression quotes are now handled separately from quasiquotes and their contents parsed properly. Fixes #116.

* Explicit brace layouts are now handled correctly. Fixes #92.

* Function application with multiple block arguments is handled correctly.
  Example:
  ```haskell
  a = do \ a -> a
      case a of
        a -> a
      do a + a
      if
     | a -> a
     | a -> a
    *
    do a
  ```

* Unicode categories for identifiers now match GHC, and the full unicode character set is supported for things like prefix operator detection.

* Haddock comments have dedicated nodes now.

* Use named precedences instead of closely replicating the GHC parser's productions.

* Different layouts are tracked and closed with their special cases considered. In particular, multi-way if now has layout.

* Fixed CPP bug where mid-line `#endif` would be false positive.

* CPP only matches legal directives now.

* Generally more lenient parsing than GHC, and in the presence of errors:
  * Missing closing tokens at EOF are tolerated for:
  	* CPP
  	* Comment
  	* TH Quotation
  * Multiple semicolons in some positions like `if/then`
  * Unboxed tuples and sums are allowed to have arbitrary numbers of filled positions

* List comprehensions can have multiple sets of qualifiers (`ParallelListComp`).

* Deriving clauses after GADTs don't require layout anymore.

* Newtype instance heads are working properly now.

* Escaping newlines in CPP works now.

* ~~One remaining issue is that qualified left sections that contain infix ops are broken: `(a + a A.+)` I haven't managed to figure out a good strategy for this – my suspicion is that it's impossible to correctly parse application, infix and negation without lexing all qualified names in the scanner. I will try that out at some point, but for now I'm planning to just accept that this one thing doesn't work. For what it's worth, none of the codebases I use for testing contain this construct in a way that breaks parsing.~~ Solved this by implementing qualified operator lookahead.

* Repo now includes a Haskell program that generates C code for classifying characters as belonging to some sets of Unicode categories, using bitmaps. I might need to change this to write them all to a shared file, so the set of source files stays the same.

### CODE REVIEW

**Comment by clason** [comment]
> I would also prefer to change the workflow so that parser.c is only generated on a release branch, not for
every commit, like some other grammars do it.

In this case, please at least commit the `grammar.json`, which is much smaller and has meaningful diffs.

Important question: will queries remain compatible?

**Comment by tek** [comment]
> In this case, please at least commit the `grammar.json`, which is much smaller and has meaningful diffs. This will allow people to generate the parser without having to use node.

Ah, didn't know that. How do you generate the parser from that file?

> You can use `tree-sitter build --wasm` with the latest version. Please upload that as a release artifact as well.

That's how we're building it, but @wenkokke added a patch for `web-tree-sitter` for some reason I don't recall.

I regenerated the `.wasm` file and included it in the PR. Are you talking about Github release artifacts?

> Important question: will queries remain compatible?

Some of them will, but it's gonna depend on the result of the survey how contained the breakage will be.
I think most expression queries will still work, but types and top level structures should be affected.

**Comment by clason** [comment]
> Ah, didn't know that. How do you generate the parser from that file?

Just `tree-sitter generate grammar.json` (little known fact).

> That's how we're building it, but @wenkokke added a patch for web-tree-sitter for some reason I don't recall.

Possibly because the output dir has changed? But that should be fixed upstream.

> I regenerated the .wasm file and included it in the PR. Are you talking about Github release artifacts?

Yes, the release artifacts (only; it shouldn't be committed to the repo).

> Some of them will, but it's gonna depend on the result of the survey how contained the breakage will be.
I think most expression queries will still work, but types and top level structures should be affected.

Thanks for the heads-up. I hope the queries here can serve as a guide for nvim-treesitter.

**Comment by 414owen** [comment]
Wow @tek this sounds incredible, clearly a labour of love, and I'm hugely grateful that you undertook such a huge amount of work for the community. The sheer quantity of listed improvements is astounding.

I'll try my best to look over it, but given the large amount of grammar changes, and that I've mostly just dealt with the scanner in the past, you're clearly the one who needs to decide when this is ready :)

🚀 🚀 🚀

File: test/corpus/prec.txt
Comment (414owen) [inline test/corpus/prec.txt:327]: Very nice.

Next step is to fix this in GHC's parser, yeah?

```
λ> [False..True]

<interactive>:1:2: error:
    Not in scope: ‘False..’
```

**Comment by wenkokke** [comment]
> That's how we're building it, but @wenkokke added a patch for `web-tree-sitter` for some reason I don't recall.

The reason is that web-tree-sitter doesn't support all of libc, and there's a file in the web-tree-sitter source that determines exactly what symbols are supported.

Please test the WASM build before merging. Tests using only the standard testing infrastructure sadly aren't reliable.

**Comment by clason** [comment]
That is an upstream issue, surely? The point of wasm is to be independent of both the source and the consumer. If something isn't working that should be working, tree-sitter should know about it rather than band-aiding it?

**Comment by tek** [comment]
@414owen thanks so much! I definitely went a bit overboard with the obsession over the last few weeks 😁 
I don't think it's necessary to review the scanner in detail – running that fuzzer should be enough to complement the tests.

File: test/corpus/prec.txt
Comment (tek) [reply test/corpus/prec.txt:327]: ah 😅 GHC makes up for it by permitting `..` to be referred to as an operator if you use a qualifying module, like `a A... a` 😁

**Comment by tek** [comment]
> Please test the WASM build before merging. Tests using only the standard testing infrastructure sadly aren't reliable.

Ah, I'll make sure to bring the wasm build and release up to speed then

**Comment by tek** [comment]
FYI @wenkokke it appears that the file with the exports was refactored in https://github.com/tree-sitter/tree-sitter/commit/d2900510f61b814e71ab5e3e0d8b59be2d276a6e, now not containing standard library symbols anymore

**Review by 414owen — APPROVED** [review]

File: src/scanner.c
Comment (tek) [reply src/scanner.c:507]: What would you consider a better name?

It's returning the nth lookahead character without advancing over it, _if_ it hasn't been advanced over before.
I think it's possible that I've rearranged the steps in a way that the second part never happens, by now.

**Comment by wenkokke** [comment]
> That is an upstream issue, surely? The point of wasm is to be independent of both the source and the consumer. If something isn't working that should be working, tree-sitter should know about it rather than band-aiding it?

This isn't a WASM issue, but rather an emscripten feature that lets you not ship the entirety of libc and libc++ for your web app.

That means that tree-sitter, as the entry point, bundles libc and libc++ and must pick what parts they want. They generally pick only those parts they need for there library itself to run.

**Comment by wenkokke** [comment]
> FYI @wenkokke it appears that the file with the exports was refactored in https://github.com/tree-sitter/tree-sitter/commit/d2900510f61b814e71ab5e3e0d8b59be2d276a6e, now not containing standard library symbols anymore
> 
> 
> 
> edit: Ah! It's in [here](https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/wasm/stdlib-symbols.txt) now, and appears to contain `realloc`, which your patch added. So we're good?

Probably, but you've rewritten the C code, so if you depend on anything new we might need the infrastructure again. Though it's more advisable to just remove functions that aren't supported.

**Comment by clason** [comment]
> They generally pick only those parts they need for there library itself to run.

Not true; they also expose (more and more) for external scanners to use. Hence my recommendation to raise an upstream issue.

**Comment by mrcjkb** [comment]
> Important question: will queries remain compatible?

@clason Since I've worked on a lot of the queries in nvim-treesitter, and also maintain [neotest-haskell](https://github.com/mrcjkb/neotest-haskell), which also uses tree-sitter queries, I'll be taking part in the survey.
Personally, I'm quite excited about some of the proposed changes and wouldn't mind backward compatibility breaking. And of course, I'd be happy to adjust the queries in nvim-treesitter.

File: src/scanner.c
Comment (414owen) [inline src/scanner.c:482]: I'm not sure why we're advancing with relation to the peek position, and I'm not convinced this for look is right anyway. Shouldn't the length of the lookahead be a limit?

**Comment by clason** [comment]
@mrcjkb Thank you, then I don't need to worry :) (To be clear, I didn't want to argue for _not_ changing the queries; I just wanted a heads-up how much breakage I should expect so I can brace for it.)

File: src/scanner.c
Comment (414owen) [reply src/scanner.c:507]: Well, it looks like `unsafe_peek_abs` implements a pure peek with `n` lookahead, and I'm not entirely sure what `peek` does. It has a few layers, and I don't understand them. They seem to try to advance in relation to a parameter and the lookahead offset?

**Comment by tek** [comment]
> Probably, but you've rewritten the C code, so if you depend on anything new we might need the infrastructure again. Though it's more advisable to just remove functions that aren't supported.

@wenkokke I fiddled with the test runner to get it to work for wasm, and it appears that everything is working – when I run `test/parse-libs wasm`, all libraries are parsed successfully!

**Comment by tek** [comment]
@414owen I'll add comments to the scanner later, so you won't have to keep guessing 😄

File: src/scanner.c
Comment (tek) [reply src/scanner.c]: yep that works!

File: src/scanner.c
Comment (tek) [reply src/scanner.c]: The docs now explain it, but for the record:

`lookahead` is the range that's _already_ been advanced over, so this starts at its end.
`offset` is the reference point for indexing `lookahead` that can be reset at any point for convenience, so the target index is `offset + n`.
If that is smaller than `lookahead`'s current size, the target index has already been consumed and the loop immediately returns.

File: src/scanner.c
Comment (tek) [reply src/scanner.c:669]: for the record:

`peek(n)` := return lookahead character `offset + n` without advancing over it, but advance over all preceding characters if necessary (may have advanced over `offset + n` before, in which case it is returned directly from our buffer)
`peek_unsafe(n)` := return lookahead character `offset + n` without advancing, return 0 if lookahead buffer is too small
`peek_unsafe_abs` := return lookahead character `n` without advancing, return 0 if lookahead buffer is too small

**Comment by tek** [comment]
@amaanq can you tell me whether the scanner is ever executed concurrently? I've made the state pointer a global variable that I assign in `tree_sitter_haskell_external_scanner_scan` to avoid having it as a parameter of every function, so it's not thread safe.

File: test/parse-lib
Comment (clason) [inline test/parse-lib:41]: it's `tree-sitter build --wasm -o ...` now.

Comment (tek) [reply test/parse-lib:41]: thanks. 0.22 doesn't appear to be available from npm or nixpkgs yet though

**Comment by wenkokke** [comment]
> @amaanq can you tell me whether the scanner is ever executed concurrently? I've made the state pointer a global variable that I assign in `tree_sitter_haskell_external_scanner_scan` to avoid having it as a parameter of every function, so it's not thread safe.

Would this result in problems if a single instance of the library is used, e.g., for multiple tabs in VSCode?

**Comment by tek** [comment]
> Would this result in problems if the library is used, e.g., to parse the contents of multiple tabs in VSCode concurrently? I think it likely will.

I don't see any obvious uses of mutexes in https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/parser.c#L437, so I guess it's up to the editor ☹️ 

ah well, probably easier to just add all the function arguments back. if we could just use C11 those pointers could get thread-local storage class... 🙄

**Comment by 414owen** [comment]
> ah well, probably easier to just add all the function arguments back. if we could just use C11 those pointers could get thread-local storage class... 🙄

I support passing the state along explicitly.
* There are [real performance concerns](https://stackoverflow.com/questions/9909980/how-fast-is-thread-local-variable-access-on-linux) to be taken into account.
* There are [real portability concerns](https://gcc.gnu.org/onlinedocs/gcc/Thread-Local.html) that may or may not be relevant anymore, but I'd rather not find out.
  >  It requires significant support from the linker (ld), dynamic linker (ld.so), and system libraries (libc.so and libpthread.so), so it is not available everywhere
* Passing around a state struct pointer explicitly feels idiomatic, and clear.

File: src/scanner.c
Comment (414owen) [reply src/scanner.c]: Okay, so there's an internal state, and a tree-sitter state.

`advance`, now pushes the current token to a the internal-state lookahead, and advances the tree-sitter character.

`peek` peeks into a internal lookahead buffer, and may advance the tree-sitter state to fill that buffer, if necessary.

So when parsing a file from scratch, does that mean that the entire file is pushed character-by-character into the lookahead buffer?

By the looks of it `reset_lookahead` doesn't reset the lookahead, so much as it advances to the end of the lookahead buffer. Maybe there's a better name for this operation?

---

For my own interest, do you think it's possible for the scanner to work with only one character of lookahead?

Comment (tek) [reply src/scanner.c]: > Okay, so there's an internal state, and a tree-sitter state.
> 
> `advance`, now pushes the current token to a the internal-state lookahead, and advances the tree-sitter character.
> 
> `peek` peeks into a internal lookahead buffer, and may advance the tree-sitter state to fill that buffer, if necessary.
> 
> So when parsing a file from scratch, does that mean that the entire file is pushed character-by-character into the lookahead buffer?

If the question is whether the entire file is stored in that buffer _entirely_ at some point, then no – the buffer is discarded after each run.
Right now the array is reused across runs and only overwritten to achieve that, but like the global pointers this isn't thread safe, so I'll have to revert to the initial approach I used and allocate a new one in each run (unlike the state pointers, this would only be unsafe if multiple nodes _in the same file_ would be processed concurrently, so it might still be feasible if we knew that TS doesn't allow that).

It's probably useful to point out that none of this has any impact on performance.
I've never observed any differences in benchmarks when changing implementation details of the scanner on a "technical" level, only when modifying logic so that the _number_ of times the scanner is called varies.
(On that note, moving stuff from the grammar to the scanner usually results in _significant_ perf benefits 😅 e.g. when I moved pragmas in, HLS parse time reduced from 420ms to 360ms)

I've also looked at memory usage but the profiling capabilities are limited with simple tools.
I only measured RSS, which varied between 19MB and 21MB for `ghc/compiler` before and after rewrite.

> By the looks of it `reset_lookahead` doesn't reset the lookahead, so much as it advances the internal state to the end of the lookahead buffer. Maybe there's a better name for this operation?

I don't have a clear idea of what "resetting lookahead" canonically means, so if you feel confident about an alternative I'm happy to change it 🙂
My interpretation is that `peek0()` is the analogue of `env->lexer->lookahead`, and resetting the offset makes the former return the same as the latter.

> For my own interest, do you think it's possible for the scanner to work with only one character of lookahead?

I think it's possible that some things would need to be spread over two runs like newline lookahead, but you could probably implement it in a convoluted way.
The main motivation is to be able to write the individual fragments independently though, without having to pay attention to their order.

**Comment by tek** [comment]
@clason 

> Just `tree-sitter generate src/grammar.json` (little known fact).

If you cloned the repo and used `tree-sitter generate`, would it make much of a difference whether the input was the javascript grammar or `grammar.json`?

> Yes, the Github release artifacts (only; it shouldn't be committed to the repo).

I added a CI job that publishes the wasm artifact to the releases.

FYI @mrcjkb I copied the queries from nvim-treesitter and adapted them, if you want to give them a once-over.

**Comment by clason** [comment]
> If you cloned the repo and used tree-sitter generate, would it make much of a difference whether the input was the javascript grammar or grammar.json?

Depends; for some grammars it is. The big benefit is if the javascript needs to be interpreted with `npm` (e.g., if you `import` other grammars); this won't be needed if you generate from the `json`.

**Comment by tek** [comment]
> Depends; for some grammars it is. The big benefit is if the javascript needs to be interpreted with `npm` (e.g., if you `import` other grammars); this won't be needed if you generate from the `json`.

ah, interesting

**Comment by amaanq** [comment]
> @amaanq can you tell me whether the scanner is ever executed concurrently? I've made the state pointer a global variable that I assign in `tree_sitter_haskell_external_scanner_scan` to avoid having it as a parameter of every function, so it's not thread safe.

Do **not** do this, it *can* be executed concurrently in the sense that two different parses might be running, which will affect this global variable, and TLS is also *not* recommended at all, for reasons Owen mentioned.

**Comment by tek** [comment]
@amaanq

> it _can_ be executed concurrently in the sense that two different parses might be running

thanks! Is it impossible for two threads to concurrently run the scanner on the same tree (i.e. same serialized state)?

**Comment by tek** [comment]
hey @wenkokke I posted a question for you on the discourse, in case you don't check your notifications there ☺️

**Comment by wenkokke** [comment]
> hey @wenkokke I posted a question for you on the discourse, in case you don't check your notifications there ☺️ 

I extremely don't, thanks for the heads up.

**Comment by wenkokke** [comment]
Unrelated point, but using field names really really simplifies queries. Does the new grammar use field names?

**Comment by tek** [comment]
> Unrelated point, but using field names really really simplifies queries. Does the new grammar use field names?

I have already added many and plan to provide full coverage!

If you have any opinions about naming or some aspect I don't know about, please let me know

**Comment by clason** [comment]
Friendly reminder to keep the `grammar.json` in the repo, or nvim-treesitter will not be able to install this parser.

**Comment by tek** [comment]
> Friendly reminder to keep the `grammar.json` in the repo, or nvim-treesitter will not be able to install this parser.

yep on my list 😅 I'm postponing this till the end to avoid cluttering my dev history

**Comment by tek** [comment]
alright, I think we can merge tomorrow if there's nothing else

**Comment by tek** [comment]
here goes nothing

**Comment by clason** [comment]
So this was merged to the `main` branch -- will that become the default branch at some point?

**Comment by tek** [comment]
@clason not sure – for now I would leave `master` as default so that `parser.c` is present when the repo is cloned. If you have suggestions for a good system I'm all ears.

(in case it's unclear – the `release` workflow will generate `parser.c` and create a commit on `master` whenever a tag is pushed)

**Comment by clason** [comment]
Ah, `master` is your "release branch". Some parsers do that, but I never understood the point. Isn't it easier to have a release workflow that makes an actual release with artifacts (tarball and wasm) for downloading, and keep the repo as "pure source"?

**Comment by tek** [comment]
I would prefer that, but as far as I could tell there are consumers who want to have a `parser.c` in the source tree.

(I've also created artifact release workflows btw.)

**Comment by clason** [comment]
> I would prefer that, but as far as I could tell there are consumers who want to have a parser.c in the source 

Yes, but does it have to be in a git branch? You can just generate the parser.c in-tree and then zip the whole thing up; the output will be identical from cloning the repo (ignoring the `.git` folder).

**Comment by tek** [comment]
my motivation was that if someone uses the plain github URL in their application to pull the repo, they'll get the generated parser like it used to be the case in the past.

When I researched this issue I got the impression that some projects depend on this, but if you have better insight into the ecosystem I'd be happy to scrap that workflow.

**Comment by clason** [comment]
Well, I understand not wanting to make sudden breaking changes on `master`, that's completely fine. I am just honestly curious why people depend on this specific setup. _If_ you are only committing the parser to `master` on manual tags/releases, then there should be zero difference from a downstream pov between an `archive/<tag>.tar.gz` link and a `releases/download/<tag>/tree-sitter-haskell.tar.gz` link?

(Unless I misunderstood and you are pushing on every commit to `main` -- but then you're back at square one with `master` bloating?)

**Comment by tek** [comment]
no, you understood that fine. I don't like it either, but that's what I learned 🤷🏻 I saw that tree-sitter-swift had multiple [requests](https://github.com/alex-pinkus/tree-sitter-swift/issues/338) to include the generated parser, even though they have a separate release branch! so my reasoning was that keeping `master` as a legacy branch will avoid having to respond to one of those once a month 😅 at least until it has become the new standard.

Of course you could argue that not playing that game would accelerate adoption...

**Comment by amaanq** [comment]
You should just check in parser.c into master for the time being, and not have a "main" branch + "release" branch. In the future, we will be getting rid of the need for this but for now, this is the way to go. I also wish you had waited till I checked this out - I'm not too familiar with Haskell so I wouldn't be particularly useful in commenting on the grammar, but I would have like to checked changes around everything else in this PR.

**Comment by clason** [comment]
I could understand that if people wanted to pull in that branch as a git submodule (in which case 🤢)...

I do think the best strategy is "keep parser.c out of repo (unless it's small), and create release artefacts for 1. the full repo tree **plus** the generated parser.c and 2. the generated wasm". Downstream users who can't handle generating their own parser shouldn't YOLO track master, either.

**Comment by tek** [comment]
@amaanq no reason why you shouldn't be able to check the changes – nothing has been released yet.

@clason understandable or not, people will open issues without reading the docs. I'd rather avoid having to deal with it over and over again

**Comment by clason** [comment]
People will open issues without reading the docs no matter what you do ;)

But I am not trying to convince you of anything; I just wanted to know the plan so nvim-treesitter can adapt. In this case, we won't have to do anything for a long while since we track the default branch.

**Comment by tek** [comment]
> People will open issues without reading the docs no matter what you do ;)

well they certainly won't open any about `parser.c` missing in the default branch if I keep it there!

> But I am not trying to convince you of anything; I just wanted to know the plan so nvim-treesitter can adapt. In this case, we won't have to do anything for a long while since we track the default branch.

I would have expected you to be eager to consume the tags or artifacts instead, irrespective of the existence of a legacy branch.

**Comment by clason** [comment]
We are not set up for that yet (because the ecosystem isn't) but that is definitely where we want to be going.

**Comment by tek** [comment]
oh I see. well in any case I'd assume that at some point all repos will be recreated from scratch in the `tree-sitter-grammars` org or something, without committing generated files going forward

**Comment by tek** [comment]
hrm, somehow rebasing broke the working tree 🤔

**Comment by clason** [comment]
> oh I see. well in any case I'd assume that at some point all repos will be recreated from scratch in the tree-sitter-grammars org or something, without committing generated files going forward

Not really, time is finite and the number of parsers is seemingly infinite. The org was just offering a home to a number of parsers looking for help with automated maintenance.

But, yes, the upstream guidance is to omit parser.c (but not grammar.json) from the repo in favor of versioned release artifacts with generated files, and we are very much looking forward to being able to rely on the latter more widely. Upstream is actively working on adding tooling to make this easier. 

For now, my curiosity is satisfied that there will not be a breaking update to the master branch (which we track and so any breaking changes there _force_ us to act) anytime soon, so I can wait and see how things play out. If anybody jumps the gun and makes a PR switching the branch, I'll now know what's what.

**Pull request merged**
